abi.go

```diff
@@ -0,0 +1,446 @@
+ * limitations under the License.
+ *
+ */
+
+package hedera
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"io"
+	"regexp"
+	"strings"
+	"sync"
+
+	"golang.org/x/crypto/sha3"
+)
+
+// ABI represents the ethereum abi format
+type ABI struct {
+	Constructor        *Method
+	Methods            map[string]*Method
+	MethodsBySignature map[string]*Method
+	Events             map[string]*Event
+	Errors             map[string]*Error
+}
+
+func (a *ABI) GetMethod(name string) *Method {
+	m := a.Methods[name]
+	return m
+}
+
+func (a *ABI) GetMethodBySignature(methodSignature string) *Method {
+	m := a.MethodsBySignature[methodSignature]
+	return m
+}
+
+func (a *ABI) addError(e *Error) {
+	if len(a.Errors) == 0 {
+		a.Errors = map[string]*Error{}
+	}
+	a.Errors[e.Name] = e
+}
+
+func (a *ABI) addEvent(e *Event) {
+	if len(a.Events) == 0 {
+		a.Events = map[string]*Event{}
+	}
+	name := overloadedName(e.Name, func(s string) bool {
+		_, ok := a.Events[s]
+		return ok
+	})
+	a.Events[name] = e
+}
+
+func (a *ABI) addMethod(m *Method) {
+	if len(a.Methods) == 0 {
+		a.Methods = map[string]*Method{}
+	}
+	if len(a.MethodsBySignature) == 0 {
+		a.MethodsBySignature = map[string]*Method{}
+	}
+	name := overloadedName(m.Name, func(s string) bool {
+		_, ok := a.Methods[s]
+		return ok
+	})
+	a.Methods[name] = m
+	a.MethodsBySignature[m.Sig()] = m
+}
+
+func overloadedName(rawName string, isAvail func(string) bool) string {
+	name := rawName
+	ok := isAvail(name)
+	for idx := 0; ok; idx++ {
+		name = fmt.Sprintf("%s%d", rawName, idx)
+		ok = isAvail(name)
+	}
+	return name
+}
+
+// NewABI returns a parsed ABI struct
+func NewABI(s string) (*ABI, error) {
+	return NewABIFromReader(bytes.NewReader([]byte(s)))
+}
+
+// MustNewABI returns a parsed ABI contract or panics if fails
+func MustNewABI(s string) *ABI {
+	a, err := NewABI(s)
+	if err != nil {
+		panic(err)
+	}
+	return a
+}
+
+// NewABIFromReader returns an ABI object from a reader
+func NewABIFromReader(r io.Reader) (*ABI, error) {
+	var abi *ABI
+	dec := json.NewDecoder(r)
+	if err := dec.Decode(&abi); err != nil {
+		return nil, err
+	}
+	return abi, nil
+}
+
+// UnmarshalJSON implements json.Unmarshaler interface
+// nolint
+func (a *ABI) UnmarshalJSON(data []byte) error {
+	var fields []struct {
+		Type            string
+		Name            string
+		Constant        bool
+		Anonymous       bool
+		StateMutability string
+		Inputs          []*ArgumentStr
+		Outputs         []*ArgumentStr
+	}
+
+	if err := json.Unmarshal(data, &fields); err != nil {
+		return err
+	}
+
+	for _, field := range fields {
+		switch field.Type {
+		case "constructor":
+			if a.Constructor != nil {
+				return fmt.Errorf("multiple constructor declaration")
+			}
+			input, err := NewTupleTypeFromArgs(field.Inputs)
+			if err != nil {
+				panic(err)
+			}
+			a.Constructor = &Method{
+				Inputs: input,
+			}
+
+		case "function", "":
+			c := field.Constant
```

Comment:
Do we need `field.Constant` if it is just used to assign `c` and `c` is always the result of `field.StateMutability == "view" || field.StateMutability == "pure"`